pkg/sql/virtual_schema.go

```diff
@@ -290,9 +361,127 @@ func (e virtualDefEntry) getPlanInfo() (sqlbase.ResultColumns, virtualTableConst
 							col.Name, col.Typ, datum.ResolvedType())
 					}
 				}
-				_, err := v.rows.AddRow(ctx, datums)
+			}
+			if !constrainedScan {
+				// If we don't have a constrained scan, just populate all the rows and
+				// we're done.
+				addRow := func(datums ...tree.Datum) error {
+					validateRow(datums...)
+					_, err := v.rows.AddRow(ctx, datums)
+					return err
+				}
+				if err := def.populate(ctx, p, dbDesc, addRow); err != nil {
+					v.Close(ctx)
+					return nil, err
+				}
+				return v, nil
+			}
+
+			// We are now dealing with a constrained virtual index scan.
+
+			if index.ID == 1 {
+				return nil, errors.AssertionFailedf(
+					"programming error: can't constrain scan on primary virtual index of table %s", e.desc.Name)
+			}
+
+			// Figure out the ordinal position of the column that we're filtering on.
+			// TODO(jordan): there is probably a better way to do this.
+			constraintColumnOrd := table.ColumnIdxMap()[index.ColumnIDs[0]]
+			addRowIfPassesFilter := func(datums ...tree.Datum) error {
+				validateRow(datums...)
+				matchedFilter := false
+				// If we have a constraint, apply it as a filter. If the filter passes,
+				// or we had no filter at all, add the datums to the row container.
+				datumToFilter := datums[constraintColumnOrd]
+				for i := 0; i < idxConstraint.Spans.Count(); i++ {
+					span := idxConstraint.Spans.Get(i)
+					if span.HasSingleKey(p.EvalContext()) {
+						matchedFilter = datumToFilter.Compare(p.EvalContext(), span.StartKey().Value(0)) == 0
```

Comment:
Is there not an existing helper to see if a datum lives within a span? If not it might be helpful to add and move this out of here.